File: 2751. Robot Collisions.java
class Solution {
    public List<Integer> survivedRobotsHealths(int[] positions, int[] healths, String directions) {
        List<Integer> ans = new ArrayList<Integer>();
        Stack<Triplet> stack = new Stack<>();
        Triplet[] arr = new Triplet[positions.length];
        for(int i=0;i<positions.length;i++){
            arr[i]= new Triplet(positions[i],healths[i],directions.charAt(i),i);
        }
        Arrays.sort(arr,new Comparator<Triplet>(){
            public int compare(Triplet a , Triplet b){
                return a.position - b.position;
            }
        });
        for(int i =0;i<positions.length;i++){
            if(stack.isEmpty()){
                stack.add(arr[i]);
            }
            else{
                while(i<positions.length && (stack.peek().direction==arr[i].direction || (stack.peek().direction=='L' && arr[i].direction=='R') )){
                    stack.add(arr[i]);
                    i++;
                }
                if(i==positions.length)
                    break;
                boolean val = false;
                while(stack.size()>0 && stack.peek().direction=='R' && arr[i].direction=='L'){
                    Triplet x = stack.pop();
                    val = false;
                    if(x.health==arr[i].health){
                        healths[x.index]=0;
                        healths[arr[i].index]=0;
                        break;
                    }
                    else if(x.health>arr[i].health){
                        x.health--;
                        healths[x.index]--;
                        healths[arr[i].index]=0;
                        stack.add(x);
                        break;
                    }
                    else{
                        healths[arr[i].index]--;
                        healths[x.index]=0;
                        arr[i].health--;
                        val = true;
                        continue;
                    }
                }
                if(val){
                    stack.add(arr[i]);
                }
               
            }
        }
        for(int i =0;i<healths.length;i++){
            if(healths[i]!=0)
            ans.add(healths[i]);
        }
        return ans;

    }
}
class Triplet{
    int position;
    int health;
    char direction;
    int index;
    Triplet(int position, int health, char direction,int index){
        this.position = position;
        this.health=health;
        this.direction = direction;
        this.index=index;
    }
}

File: 2751. Robot Collisions.rs
impl Solution {
    pub fn survived_robots_healths(positions: Vec<i32>, mut healths: Vec<i32>, directions: String) -> Vec<i32> {
        let n = positions.len();
        let dirs: Vec<u8> = directions.bytes().collect();
        let mut order: Vec<usize> = (0..n).collect();
        order.sort_unstable_by_key(|&i| positions[i]);

        let mut stack: Vec<usize> = vec![]; // indices of 'R' moving robots (by position)

        for &i in &order {
            if dirs[i] == b'R' {
                stack.push(i);
            } else {
                // 'L' robot collides with rightmost 'R' robots
                while let Some(&j) = stack.last() {
                    if healths[j] < healths[i] {
                        healths[j] = 0;
                        healths[i] -= 1;
                        stack.pop();
                    } else if healths[j] > healths[i] {
                        healths[j] -= 1;
                        healths[i] = 0;
                        break;
                    } else {
                        healths[j] = 0;
                        healths[i] = 0;
                        stack.pop();
                        break;
                    }
                }
            }
        }
        (0..n).filter(|&i| healths[i] > 0).map(|i| healths[i]).collect()
    }
}
